### ISSUE TITLE
[2.3.1] - DynamicRecord Not in use, blockId[2]

### ISSUE THREAD

**Issue by clun** [neo4j/neo4j#6157]
Dear all,

I about to read nodes properties from an embedded database and got the following exception.

```
org.neo4j.kernel.impl.store.InvalidRecordException: DynamicRecord Not in use, blockId[2]
    at org.neo4j.kernel.impl.store.AbstractDynamicStore.checkForInUse(AbstractDynamicStore.java:450)
    at org.neo4j.kernel.impl.store.AbstractDynamicStore.access$300(AbstractDynamicStore.java:66)
    at org.neo4j.kernel.impl.store.AbstractDynamicStore$DynamicRecordCursor.next(AbstractDynamicStore.java:692)
    at org.neo4j.kernel.impl.api.store.StorePropertyPayloadCursor.readFromStore(StorePropertyPayloadCursor.java:268)
    at org.neo4j.kernel.impl.api.store.StorePropertyPayloadCursor.stringValue(StorePropertyPayloadCursor.java:205)
    at org.neo4j.kernel.impl.api.store.StorePropertyCursor.payloadValueAsObject(StorePropertyCursor.java:155)
    at org.neo4j.kernel.impl.api.store.StorePropertyCursor.value(StorePropertyCursor.java:109)
    at org.neo4j.kernel.api.cursor.EntityItem$EntityItemHelper.getProperty(EntityItem.java:51)
    at org.neo4j.kernel.impl.api.OperationsFacade.nodeGetProperty(OperationsFacade.java:308)
    at org.neo4j.kernel.impl.core.NodeProxy.getProperty(NodeProxy.java:472)
    at org.ff4j.store.FeatureStoreNeo4J.read(FeatureStoreNeo4J.java:151)
    at org.ff4j.neo4j.DefaultUnitTesting.getFeatureByUId(DefaultUnitTesting.java:111)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.lang.reflect.Method.invoke(Unknown Source)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

It's a bit strange, there are 2 properties on the sample node I can read the first one (named initParams) but not the second. (named class).

The request performed is : 

```
MATCH (f:FF4J_FEATURE { uid: 'first' })--(all) RETURN f,all
```

In the browser I successfully display the result (no more embedded there of course)
![neo4j-1](https://cloud.githubusercontent.com/assets/726536/11954413/96a9b9aa-a8a8-11e5-80d6-1698deb3517e.png)

![neo-2](https://cloud.githubusercontent.com/assets/726536/11954417/9d3c53fe-a8a8-11e5-99d1-9f052ad765d8.png)

But when reading from Java I got the Issue. Here is the snippied

``` java
try (Transaction tx = graphDb.beginTx()) {

            Result result = graphDb.execute(QUERY_CYPHER_READ_FEATURE,  queryParameters);

            // Parsing
            while (result.hasNext()) {
                Map < String, Object > response = result.next();

                // Map for 'all'
                Node   nodeOther = (Node) response.get("all");
                String nodeLabel = nodeOther.getLabels().iterator().next().toString();
                FF4jNeo4jLabels lblb = FF4jNeo4jLabels.valueOf(nodeLabel);
                switch(lblb) {
                    case FF4J_FLIPPING_STRATEGY:
                        System.out.println("FF4J_FLIPPING_STRATEGY");

                        // This line works
                        nodeOther.getProperty("initParams");

                        // This line failed
                        nodeOther.getProperty("clazz");
                    break;

                }
            }
            tx.success();
        }
```

You can find the whole sample here (it's for FF4J open source framework : 

https://github.com/clun/ff4j/blob/master/ff4j-store-neo4j/src/test/java/org/ff4j/neo4j/DefaultUnitTesting.java

https://github.com/clun/ff4j/blob/master/ff4j-store-neo4j/src/main/java/org/ff4j/store/FeatureStoreNeo4J.java

**Comment by clun** [comment]
After several tests I notice that the problem occurs when and only when the value length is more than 36 characters.

**Comment by chrisvest** [comment]
Hi @clun 

The reason is that you are setting different array and string block sizes:
https://github.com/clun/ff4j/blob/master/ff4j-store-neo4j/src/test/java/org/ff4j/neo4j/DefaultUnitTesting.java#L31-L34
And we have a bug where we can end up using the wrong setting, e.g. use the array block size where we should have used the string block size.

This will be fixed by #6141
